### PR TITLE
Day13

### PR DESCRIPTION
**Pull request by KoenvdBerg**

### CODE REVIEW

File: aoc2018/src/main/scala/day13.scala [hunk 1]
```diff
@@ -1,8 +1,144 @@
...
+
+  def rotateVector[A](n: Int, s: Vector[A]): Vector[A] =
+    if s.isEmpty then s
+    else
+      val nbound: Int = n % s.length // skipping the full rotation rounds
+      if nbound < 0 then rotateVector(nbound + s.length, s)
+      else s.drop(nbound) ++ s.take(nbound)
+
+  case class Point(x: Int, y: Int):
+    def add(p2: Point): Point = Point(x + p2.x, y + p2.y)
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:23]: note that symbolic definitions are also feasible - especially for coordinate use cases like this - eg. given two points `p1` and `p2` - `def +(p2: Point): Point = Point(x + p2.x, y + p2.y)` would yield the client side syntax `val p3 = p1 + p2`

File: aoc2018/src/main/scala/day13.scala [hunk 2]
```diff
@@ -1,8 +1,144 @@
...
+  sealed trait Switch
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:24]: could be refactored in an `enum` which is more idiomatic scala 3

File: aoc2018/src/main/scala/day13.scala [hunk 3]
```diff
@@ -1,8 +1,144 @@
...
+    def update: Cart =
+      Cart(loc.add(dir), dir, intersect)
+
+    def computeSwitch(turn: Switch): Cart =
+      turn match
+        case Straight => straight.update
+        case Back => backwardSwitch.update
+        case Forward => forwardSwitch.update
+        case Intersect => handleIntersect.update
+    def forwardSwitch: Cart =
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:42]: same comment as for `update`

File: aoc2018/src/main/scala/day13.scala [hunk 4]
```diff
@@ -1,8 +1,144 @@
...
+      Cart(loc, Point(dir.y, dir.x), intersect)
+
+    def backwardSwitch: Cart =
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:45]: same comment as for `update`

File: aoc2018/src/main/scala/day13.scala [hunk 5]
```diff
@@ -1,8 +1,144 @@
...
+      Cart(loc, Point(-dir.y, -dir.x), intersect)
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:46]: could be refactored into a `def -(that: Point): Point` on the `Point` case class itself

File: aoc2018/src/main/scala/day13.scala [hunk 6]
```diff
@@ -1,8 +1,144 @@
...
+
+    def straight: Cart =
```
Comment (nmcb) [inline aoc2018/src/main/scala/day13.scala:48]: same comment as for `update`

File: aoc2018/src/main/scala/day13.scala [hunk 7]
```diff
@@ -1,8 +1,144 @@
...
+  sealed trait Switch
+  case object Straight extends Switch
+  case object Back extends Switch
+  case object Forward extends Switch
+  case object Intersect extends Switch
+
+  case class Rail(switch: Switch, loc: Point)
+  case class Cart(loc: Point, dir: Point, intersect: Vector[Switch]):
+
+    def update: Cart =
```
[the remark this answers is not in the document]
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day13.scala:33]: Ah yes, in this case I do need the `dir` to be part of the `Cart` case class for the `handleIntersect` function to work. 

However, I agree that the update function could be written without the `dir` in the case class like you mention here.

File: aoc2018/src/main/scala/day13.scala [hunk 8]
```diff
@@ -1,8 +1,144 @@
...
+    loc match
+      case Some(p) => p
+      case None => simulateTrain(newCarts.sortBy(_.loc.x), rails)
+
+
+  private val answer1 = simulateTrain(incarts, inrails)
+  println(s"Answer day $day part 1: ${answer1} [${System.currentTimeMillis - start1}ms]")
+
+
+  def simulateTrain2(carts: Vector[Cart], rails: Vector[Rail]): Cart =
```
[the remark this answers is not in the document]
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day13.scala:116]: Okay if I leave this for now, and remember it for the upcoming AOC problems?

File: aoc2018/src/main/scala/day13.scala (hunk 2 above)
Comment (KoenvdBerg) [reply aoc2018/src/main/scala/day13.scala:24]: I guess you mean something like this??

```
  enum Switch:
    case Straight, Back, Forward, Intersect
```

How can I pattern match on that?

Comment (nmcb) [reply aoc2018/src/main/scala/day13.scala:20]: eg.

```
val s: Switch = Straight
s match
  case Straight  => ...
  case Back      => ...
  case Forward   => ...
  case Intersect => ...
```

**Review by nmcb — APPROVED** [review]
lgtm :)

**Pull request merged**